app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx

```diff
@@ -0,0 +1,278 @@
+      if (cachedSearches[searchTerm]) {
+        setUsers(cachedSearches[searchTerm]);
+        return;
+      }
+
+      fetchSuggestions(searchTerm).then(({ result: fetchedUsers }) => {
+        const resultLength = Math.min(
+          fetchedUsers.length,
+          MAX_RESULTS_DISPLAYED,
+        );
+
+        const results = fetchedUsers.slice(0, resultLength);
+
+        setCachedSearches({
+          ...cachedSearches,
+          [searchTerm]: results,
+        });
+
+        setUsers(results);
+
+        // Let screen reader users know a list has populated
+        const requiresAriaLiveAnnouncement =
+          !ariaHelperText && fetchedUsers.length > 0;
+
+        if (requiresAriaLiveAnnouncement) {
+          setAriaHelperText(
+            `Mention user, ${fetchedUsers.length} results found`,
+          );
+        }
+      });
+    }
+  }, [searchTerm, fetchSuggestions, cachedSearches, ariaHelperText]);
+
+  useLayoutEffect(() => {
+    const popover = document.getElementById('mention-autocomplete-popover');
+    if (!popover) {
+      return;
+    }
+
+    const closeOnClickOutsideListener = (event) => {
+      if (!popover.contains(event.target)) {
+        // User clicked outside, reset to not searching state
+        setSearchTerm('');
+        setAriaHelperText('');
+        setUsers([]);
+      }
+    };
+
+    document.addEventListener('click', closeOnClickOutsideListener);
+
+    return () =>
+      document.removeEventListener('click', closeOnClickOutsideListener);
+  }, [searchTerm]);
+
+  useLayoutEffect(() => {
+    inputRef.current.focus();
+    inputRef.current.setSelectionRange(cursorPosition, cursorPosition - 1);
+  }, [cursorPosition]);
+
+  const handleValueChange = ({ target: { value } }) => {
+    setTextContent(value);
+    const { isUserMention, indexOfMentionStart } = getMentionWordData(
+      inputRef.current,
+    );
+
+    const { selectionStart } = inputRef.current;
+
+    if (isUserMention) {
+      // search term begins after the @ character
+      const searchTermStartPosition = indexOfMentionStart + 1;
+
+      const mentionText = value.substring(
+        searchTermStartPosition,
+        selectionStart,
+      );
+
+      const { x: cursorX, y } = getCursorXY(
+        inputRef.current,
+        indexOfMentionStart,
+      );
+      const textAreaX = inputRef.current.offsetLeft;
+
+      // On small screens always show dropdown at start of textarea
+      const dropdownX = isSmallScreen ? textAreaX : cursorX;
+
+      setDropdownPositionPoints({ x: dropdownX, y });
+      setSearchTerm(mentionText);
+      setSelectionInsertIndex(searchTermStartPosition);
+    } else if (searchTerm) {
+      // User has moved away from an in-progress @mention - clear current search
+      setSearchTerm('');
+      setAriaHelperText('');
+      setUsers([]);
+    }
+  };
+
+  const handleSelect = (username) => {
+    const textWithSelection = `${textContent.substring(
```

Comment:
Documentation (blocking): Add a short comment about what textWithSelection is building.